### ISSUE TITLE
lax.switch compilation time is super-linear in number of switch options

### ISSUE THREAD

**Issue by j-towns** [google/jax#4453]
To demonstrate this, you can run the following (results below):
```python
from time import time

import matplotlib.pyplot as plt

from jax import jit, lax
import jax.numpy as np


def time_switch(n):
  # Create list of functions
  inners = []
  for i in range(n):
    inners.append(lambda x: x ** 2)
  switch = jit(lambda i, x: lax.switch(i, inners, x))
  t0 = time()
  switch(0, 3.)
  return time() - t0

def bench_switch(ns):
  return [time_switch(n) for n in ns]

ns = [256, 512, 768, 1024, 1280, 1536, 1792, 2048]
times = bench_switch(ns)
plt.plot(ns, times)
```
The bottleneck appears to be in XLA compilation; I profiled `time_switch` (with n=1500 on CPU) and almost all time is spent in [`backend_compile`](https://github.com/google/jax/blob/51538bd76d510487c137b9bffccb1e7f9ad58173/jax/interpreters/xla.py#L341-L344):
<img width="1030" alt="Screenshot 2020-10-05 at 14 09 31" src="https://user-images.githubusercontent.com/15261883/95083427-6a5aad80-0714-11eb-8f7f-7e5ebdf02097.png">.

### Benchmark results (y-axis in seconds)
CPU
![cpu](https://user-images.githubusercontent.com/15261883/95083705-c3c2dc80-0714-11eb-93a9-9008f2c5668a.png)

GPU
![gpu](https://user-images.githubusercontent.com/15261883/95083715-c9202700-0714-11eb-8dfc-43ee321fa526.png)

TPU
![tpu](https://user-images.githubusercontent.com/15261883/95084050-40ee5180-0715-11eb-942e-7bee2ac3676c.png)

**Comment by j-towns** [comment]
This isn't a critical issue for me, and I think it can be worked around by using nested switches, which seem to compile faster, but I thought it was worth raising nevertheless.

**Comment by hawkinsp** [comment]
To be honest, I'm not that surprised that compilation time is superlinear for this kind of program. Looking at a C++-level profile, the majority of time is spent inside LLVM. So I doubt there's an easy fix here.

**Comment by j-towns** [comment]
Closing since this is apparently caused by LLVM so out of our control.

**Comment by j-towns** [comment]
(and I'm not in any way blocking on it).
